### ISSUE TITLE
Random IndexOutOfRangeException RazorEngine 3.7.3

### ISSUE THREAD

**Issue by zelid** [Antaris/RazorEngine#331]
Having very strange exception on RazorEngine 3.7.3 that happens randomly in production after upgrade from RazorEngine 3.4.2

```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Generic.List`1.Add(T item)
   at RazorEngine.Compilation.CrossAppDomainCleanUp.CleanupHelper.RegisterCleanupPath(String path)
   at RazorEngine.Compilation.ExecutionContextLessThread.ActionConv.Call(Boolean data)
   at RazorEngine.Compilation.ExecutionContextLessThread.CallHelperSafeHelper`2.AsAction()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at RazorEngine.Compilation.ExecutionContextLessThread.DefaultCallAction(Action a)
   at RazorEngine.Templating.InvalidatingCachingProvider.CacheTemplate(ICompiledTemplate template, ITemplateKey templateKey)
   at RazorEngine.Templating.RazorEngineService.CompileAndCacheInternal(ITemplateKey key, Type modelType)
   at RazorEngine.Templating.RazorEngineService.GetCompiledTemplate(ITemplateKey key, Type modelType, Boolean compileOnCacheMiss)
   at RazorEngine.Templating.RazorEngineService.GetTemplate(ITemplateKey key, Type modelType, Object model, DynamicViewBag viewbag)
   at RazorEngine.Templating.TemplateService.GetTemplate(String razorTemplate, Object model, String name)
   at RazorEngine.Templating.TemplateService.Parse(String razorTemplate, Object model, DynamicViewBag viewBag, String cacheName)
```

What can be the reason?

**Comment by aviatrix** [comment]
Can you show us the code that causes this exception ? both view code & how you call razorengine + config, ideally a test project to test this.

**Comment by zelid** [comment]
We use RazorEngine in a custom IHttpHandler called "Website.ashx"

```
/* non-RazorEngine application business logic start */
public void ProcessRequest(HttpContext httpContext)
{

// get page template layout from database (layout contains customer string placeholders for tags like {{BLOG_POSTS_LIST}} )
var templateLayout = GetPageContentFromDb(httpContext, viewModel);

// get template tags for placeholders used in template layout
var templateTags = viewModel.DbRecords.templateTags;
// localised template tags
var localizedTags = GetLocalizedTags(templateTags, viewModel.DbRecords.templateTagTranslations);

// replace tags with content their content in template lyout
// here "template" is fully sutable to be passed to RazorEngine
var template = Regex.Replace(templateLayout, WebsiteConfigManager.TemplateTagPattern,
    m => ReplaceTemplateTag(m, localizedTags), RegexOptions.Compiled);

/* non-RazorEngine application business logic end */

// RazorEngine usage below:

var websitePageCacheKey = string.Format("WebsitePageCache_{0}_{1}", viewModel.DbRecords.WebsiteTemplateSet.Id, viewModel.WebsiteRequest.SelectedLanguageCode);

var content = Razor.Parse(template, viewModel, websitePageCacheKey);

httpContext.Response.Write(content);
httpContext.Response.End();
```

I can't provide `template` and `viewModel` here they are large and very application specific.
The error has nothing to do with template content or viewModel structure.
Everything worked fine with RazorEngine 3.4.2

When upgraded to RazorEngine 3.7.3 rendering *randomly* crashes in very unexpected situations in 5-10% of all page opens. Before upgrade everything worked fine.

Very painful because it's already in a high-loaded production and we had no errors on our test environment.

**Comment by zelid** [comment]
Maybe this should help?

http://stackoverflow.com/questions/11978064/got-index-out-of-bounds-error-on-list-add-in-c-sharp

http://stackoverflow.com/questions/7193047/multithread-webservice-acessing-and-modifying-list-object-not-threadsafe

**Comment by aviatrix** [comment]
i would suggest several things : 
1) try to migrate your code to 3.5+ style  https://antaris.github.io/RazorEngine/Upgrading.html  . I've been down this road, that might be the reason ( the old API is being deprecated).

2) try to disable temporary file locks >
https://antaris.github.io/RazorEngine/index.html#Temporary-files
https://github.com/Antaris/RazorEngine/issues/244

**Comment by zelid** [comment]
Tried config.DisableTempFileLocking = true and new API the bug still there

The only thing that really helped - revert back to the old API and do:
```
Uninstall-Package RazorEngine -Force
Install-Package RazorEngine -Version 3.4.2
```

On all projects.

The bug reason was mentioned - there is a concurrent access to non-thread safe collection `List<string> _toCleanup` you start to "feel" it when you get many page visitors per second.

**Comment by matthid** [comment]
Try adding `config.CachingProvider = new DefaultCachingProvider(t => {})` in combination with `DisableTempFileLocking` as a hotfix until I have some time to look at it.

**Comment by campersau** [comment]
I have the same issue because `_toCleanup` is not a thread safe. I had to manually put a synclock around `RegisterCleanupPath` function calls.

**Comment by campersau** [comment]
This might be a dublicate of https://github.com/Antaris/RazorEngine/issues/325

**Comment by matthid** [comment]
fixed in 3.7.4
